### PR TITLE
Implement missing assembly output functions and do not use PushString for assembly.

### PR DESCRIPTION
**Pull request by chriseth**
Without this pull request, `solc --optimize --asm` sometimes throws exceptions.

### CODE REVIEW

File: libevmasm/Assembly.cpp [hunk 1]
```diff
@@ -205,7 +205,8 @@ ostream& Assembly::streamAsm(ostream& _out, string const& _prefix, StringMap con
 	{
 		_out << _prefix << "stop" << endl;
 		for (auto const& i: m_data)
-			assertThrow(u256(i.first) < m_subs.size(), AssemblyException, "Data not yet implemented.");
```
Comment (pirapira) [inline libevmasm/Assembly.cpp:208]: What is the situation when `m_data` contains a pair whose first element is small?  If that kind of pairs should be ignored, I guess that kind of pairs should not be inserted in the first place.  Is there a possibility that we are hiding problems here?

File: libevmasm/AssemblyItem.cpp [hunk 2]
```diff
@@ -159,18 +159,25 @@ string AssemblyItem::toAssemblyText() const
...
 		break;
 	case PushString:
-		assertThrow(false, AssemblyException, "Push string assembly output not implemented.");
+		text = string("data_") + toHex(data());
 		break;
 	case PushTag:
-		assertThrow(data() < 0x10000, AssemblyException, "Sub-assembly tags not yet implemented.");
-		text = string("tag_") + to_string(size_t(data()));
+	{
+		size_t sub;
```
Comment (pirapira) [inline libevmasm/AssemblyItem.cpp]: I thought our usual policy is to initialize these always.

Comment (chriseth) [reply libevmasm/AssemblyItem.cpp]: Ah, thanks!

File: libevmasm/Assembly.cpp (hunk 1 above)
Comment (chriseth) [reply libevmasm/Assembly.cpp:208]: The first element should be a hash, so I guess it is practically impossible to do that. These problems will all go away when we transition to the new IR and bytecode generator.

File: libevmasm/AssemblyItem.cpp (hunk 2 above)
Comment (chriseth) [reply libevmasm/AssemblyItem.cpp]: Ah ok, now I see it. So strictly, it is not required, but I think it is good style to do it regardless.

**Comment by chriseth** [comment]
@pirapira ready to merge? Note that the actual representation might change in the future depending how the IR evolves. This is just to make it not throw.

File: libevmasm/Assembly.cpp (hunk 1 above)
Comment (pirapira) [reply libevmasm/Assembly.cpp:208]: I'd like to create an issue about this.  I think, what's missing is a way to distinguish data and subassembly in this array.

Comment (pirapira) [reply libevmasm/Assembly.cpp:208]: By the way, why is `m_subs.size()` relevant here?

**Comment by pirapira** [comment]
I'm wondering if it's easier to just remove the check `i.first < m_subs.size()`.  I think I heard something like, `i.first` might be small in case of subassembly, but I didn't understand the explanation.  Can `i` be a subassembly?  It does not look like so to me.

**Comment by pirapira** [comment]
If `m_data` contains two different kinds of things, I need an issue about separating these things in different containers.  Otherwise, I think the check can be removed.

**Comment by chriseth** [comment]
Issue is here: https://github.com/ethereum/solidity/issues/2130

**Review by pirapira — APPROVED** [review]
Looks good to me.

**Pull request merged**